File: MulDivWPF/MulDiv/MDCalc.cs
using System;
using System.Windows.Media;
using System.Windows;

namespace MulDiv
{
	/// <summary>
	/// MDCalc ‚ÌŠT—v‚Ìà–¾‚Å‚·B
	/// </summary>
	public class MDCalc
	{
//		private System.Drawing.Color c_area_bg;       // ŒvZ‚Ì—Ìˆæ‚Ì”wŒiF
//		private System.Drawing.Color c_text;          // ŒvZ‚Ì•¶š‚ÌF
//		private System.Drawing.Font font_normal;
//		private System.Drawing.Pen pen_text;
//		private System.Drawing.SolidBrush brush_text;
//		private int char_width = 40;
//		private int char_height = 40;
		private int x_margin = 10;
		private int y_margin = 10;
		private const float hmag = 0.6f;
		private const float vmag = 0.6f;

		private ViewSettings settings;

		public MDCalc(ViewSettings view)
		{
//			c_area_bg = System.Drawing.Color.FromArgb(255, 245, 245); // ŒvZ‚Ì—Ìˆæ‚Ì”wŒiF
//			c_text = System.Drawing.Color.FromArgb(0, 0, 0);          // ŒvZ‚Ì•¶š‚ÌF
//			font_normal = new System.Drawing.Font("Times New Roman", 40);
//			pen_text = new System.Drawing.Pen(c_text);
//			brush_text = new System.Drawing.SolidBrush(c_text);
			settings = view;
		}

		public void paint(XGraphics g, string str, int col, int row)
		{
			if ( str != "" )
			{
				if ( settings.UseImage )
				{
					int char_width = settings.ImageSettings.GetWidth("0.");
					int char_height = settings.ImageSettings.GetHeight();
					if ( str == "." )
					{
						int width = settings.ImageSettings.GetWidth(".");
						settings.ImageSettings.DrawString(str, col * char_width - width + x_margin, row * char_height + y_margin, g);
					}
					else
					{
						settings.ImageSettings.DrawString(str, col * char_width + x_margin, row * char_height + y_margin, g);
					}
				}
				else
				{
					XFont font = settings.CalcAreaFont;
					Size size = g.MeasureString("0", font);
					int char_width = (int)(size.Width * settings.CalcAreaHorizontalMagnitude * hmag);
					int char_height = (int)(size.Height * settings.CalcAreaVerticalMagnitude * vmag);
					Brush brush_text = new SolidColorBrush(settings.CalcAreaTextColor.WColor);
					if ( str == "." )
					{
						g.DrawString(str, font, brush_text, (int)((col - 0.4f) * char_width + x_margin), row * char_height + y_margin);
					}
					else
					{
						g.DrawString(str, font, brush_text, col * char_width + x_margin, row * char_height + y_margin);
					}
				}
			}
		}
	}
}

File: MulDivTS/MulDiv/MDCalc.ts
class MDCalc
{
    private settings: ViewSettings;
    private x_margin: number;
    private y_margin: number;
    private hmag: number;
    private vmag: number;

    public constructor(view_settings: ViewSettings) {
        this.settings = view_settings;

        this.x_margin = 10;
        this.y_margin = 10;
        this.hmag = 1;
        this.vmag = 1;
    }

	public paint = function(gc: CanvasRenderingContext2D, str: string, col: number, row: number): void
	{
		if ( str != "" )
		{
			if ( this.settings.UseImage )
			{
				//int char_width = settings.ImageSettings.GetWidth("0.");
				//int char_height = settings.ImageSettings.GetHeight();
				//if ( str == "." )
				//{
				//	int width = settings.ImageSettings.GetWidth(".");
				//	settings.ImageSettings.DrawString(str, col * char_width - width + x_margin, row * char_height + y_margin, g);
				//}
				//else
				//{
				//	settings.ImageSettings.DrawString(str, col * char_width + x_margin, row * char_height + y_margin, g);
				//}
			}
			else
			{
				var font = new WYFont(this.settings.CalcAreaFont, gc, false, null);
				var w = font.stringWidth("0");
				var h = font.getHeight();
				var char_width = Math.floor(w * this.settings.CalcAreaHorizontalMagnitude * this.hmag);
				var char_height = Math.floor(h * this.settings.CalcAreaVerticalMagnitude * this.vmag);
				var g = new WYGraphics(gc, this.settings.CalcAreaFont, false, null);
				g.setColor(this.settings.CalcAreaTextColor);
				if ( str == "." )
				{
					g.drawString(str, (col - 0.4) * char_width + this.x_margin, row * char_height + this.y_margin);
				}
				else
				{
					g.drawString(str, col * char_width + this.x_margin, row * char_height + this.y_margin);
				}
			}
		}
	}
}
